File: app/Models/UserSubscription.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Carbon\Carbon;

class UserSubscription extends Model
{
    use HasFactory;

    protected $fillable = [
        'user_id',
        'plan_id',
        'status',
        'starts_at',
        'ends_at',
        'trial_ends_at',
        'cancelled_at',
        'paystack_subscription_code',
        'paystack_customer_code',
        'paystack_email_token',
        'auto_renew',
        'metadata',
    ];

    protected $casts = [
        'starts_at' => 'datetime',
        'ends_at' => 'datetime',
        'trial_ends_at' => 'datetime',
        'cancelled_at' => 'datetime',
        'auto_renew' => 'boolean',
        'metadata' => 'array',
    ];

    /**
     * The user who owns the subscription
     */
    public function user()
    {
        return $this->belongsTo(User::class);
    }

    /**
     * The subscription plan
     */
    public function plan()
    {
        return $this->belongsTo(SubscriptionPlan::class, 'plan_id');
    }

    /**
     * Transactions for this subscription
     */
    public function transactions()
    {
        return $this->hasMany(SubscriptionTransaction::class, 'subscription_id');
    }

    /**
     * Check if subscription is active
     */
    public function isActive(): bool
    {
        return $this->status === 'active' &&
               ($this->ends_at === null || $this->ends_at->isFuture());
    }

    /**
     * Check if subscription is on trial
     */
    public function onTrial(): bool
    {
        return $this->status === 'trial' &&
               $this->trial_ends_at !== null &&
               $this->trial_ends_at->isFuture();
    }

    /**
     * Check if subscription is expired
     */
    public function isExpired(): bool
    {
        return $this->status === 'expired' ||
               ($this->ends_at !== null && $this->ends_at->isPast());
    }

    /**
     * Check if subscription is cancelled
     */
    public function isCancelled(): bool
    {
        return $this->status === 'cancelled';
    }

    /**
     * Cancel subscription
     */
    public function cancel(): bool
    {
        $this->status = 'cancelled';
        $this->cancelled_at = now();
        $this->auto_renew = false;
        return $this->save();
    }

    /**
     * Activate subscription
     */
    public function activate(Carbon $endsAt = null): bool
    {
        $this->status = 'active';
        $this->starts_at = now();

        if ($endsAt) {
            $this->ends_at = $endsAt;
        } else {
            // Default to 1 month
            $this->ends_at = now()->addMonth();
        }

        return $this->save();
    }

    /**
     * Renew subscription
     */
    public function renew(int $months = 1): bool
    {
        $this->ends_at = $this->ends_at ?
            $this->ends_at->addMonths($months) :
            now()->addMonths($months);

        $this->status = 'active';

        return $this->save();
    }

    /**
     * Check if subscription can be renewed
     */
    public function canRenew(): bool
    {
        return $this->auto_renew && !$this->isCancelled();
    }
}

File: mobile/lib/models/user_subscription.dart
import 'subscription_plan.dart';

class UserSubscription {
  final int id;
  final int userId;
  final int planId;
  final String status; // active, cancelled, expired, trial
  final DateTime? startsAt;
  final DateTime? endsAt;
  final DateTime? trialEndsAt;
  final DateTime? cancelledAt;
  final String? paystackSubscriptionCode;
  final String? paystackCustomerCode;
  final bool autoRenew;
  final DateTime createdAt;
  final DateTime updatedAt;

  // Relationships
  final SubscriptionPlan? plan;

  UserSubscription({
    required this.id,
    required this.userId,
    required this.planId,
    required this.status,
    this.startsAt,
    this.endsAt,
    this.trialEndsAt,
    this.cancelledAt,
    this.paystackSubscriptionCode,
    this.paystackCustomerCode,
    required this.autoRenew,
    required this.createdAt,
    required this.updatedAt,
    this.plan,
  });

  factory UserSubscription.fromJson(Map<String, dynamic> json) {
    return UserSubscription(
      id: json['id'],
      userId: json['user_id'],
      planId: json['plan_id'],
      status: json['status'],
      startsAt: json['starts_at'] != null
          ? DateTime.parse(json['starts_at'])
          : null,
      endsAt: json['ends_at'] != null
          ? DateTime.parse(json['ends_at'])
          : null,
      trialEndsAt: json['trial_ends_at'] != null
          ? DateTime.parse(json['trial_ends_at'])
          : null,
      cancelledAt: json['cancelled_at'] != null
          ? DateTime.parse(json['cancelled_at'])
          : null,
      paystackSubscriptionCode: json['paystack_subscription_code'],
      paystackCustomerCode: json['paystack_customer_code'],
      autoRenew: json['auto_renew'] ?? false,
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
      plan: json['plan'] != null
          ? SubscriptionPlan.fromJson(json['plan'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'plan_id': planId,
      'status': status,
      'starts_at': startsAt?.toIso8601String(),
      'ends_at': endsAt?.toIso8601String(),
      'trial_ends_at': trialEndsAt?.toIso8601String(),
      'cancelled_at': cancelledAt?.toIso8601String(),
      'paystack_subscription_code': paystackSubscriptionCode,
      'paystack_customer_code': paystackCustomerCode,
      'auto_renew': autoRenew,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  bool get isActive => status == 'active';
  bool get isCancelled => status == 'cancelled';
  bool get isExpired => status == 'expired';
  bool get isTrial => status == 'trial';

  bool get isExpiringSoon {
    if (endsAt == null) return false;
    final daysUntilExpiry = endsAt!.difference(DateTime.now()).inDays;
    return daysUntilExpiry <= 7 && daysUntilExpiry >= 0;
  }

  int get daysUntilExpiry {
    if (endsAt == null) return -1;
    return endsAt!.difference(DateTime.now()).inDays;
  }
}
